### ISSUE TITLE
github actions problems with test scripts starting /usr/bin/pcscd -f 

### ISSUE THREAD

**Issue by dengert** [OpenSC/OpenSC#2390]
Some of the github actions test scripts running in Ubuntu fail with 
`00000000 [140024449624000] utils.c:81:GetDaemonPid() Can't open /run/pcscd/pcscd.pid: No such file or directory`

The problems appears in the [test-piv.sh lines 26-29](https://github.com/OpenSC/OpenSC/blob/master/.github/test-piv.sh#L26-L29)
./test-isoapplet.sh, ./test-gidsapplet.sh and ./test-openpgp.sh all do the same.

This first problem is `PCSCD_PID=$!` returns the PID of the sudo command, not the /usr/sbin/pcscd -f
Because /usr/sbin/pcscd  is started by sudo, not be the shell, and sudo still running when     https://github.com/OpenSC/OpenSC/blob/master/.github/test-piv.sh#L45 kills the sudo, but not the /usr/sbin/pcscd keeps running. So after a few tests, there are multiple  "/usr/sbin/pcscd -f" running.
Since "bash -e" is used, line 45 may never be executed if there is a failure before. 

The second problem is a timing issue because /usr/sbin/pcscd takes some time to get started before it sets `/run/pcscd/pcscd.pid` so any following command may fail.

On my Ubuntu-20.04, (/var/run is symlink to /run)  and systemctl shows after stopping and starting to clean up from some testing:
```
doug@XUbuntu-20:/tmp$ sudo systemctl stop pcscd.service pcscd.socket
doug@XUbuntu-20:/tmp$ sudo systemctl start pcscd.service pcscd.socket
doug@XUbuntu-20:/tmp$ sudo systemctl status pcscd
● pcscd.service - PC/SC Smart Card Daemon
     Loaded: loaded (/lib/systemd/system/pcscd.service; indirect; vendor preset: enabled)
     Active: active (running) since Wed 2021-08-18 10:04:19 CDT; 16s ago
TriggeredBy: ● pcscd.socket
       Docs: man:pcscd(8)
   Main PID: 6646 (pcscd)
      Tasks: 3 (limit: 9485)
     Memory: 692.0K
     CGroup: /system.slice/pcscd.service
             └─6646 /usr/sbin/pcscd --foreground --auto-exit

Aug 18 10:04:19 XUbuntu-20.04 systemd[1]: Started PC/SC Smart Card Daemon.
doug@XUbuntu-20:/tmp$ ls -l /var/run/pcscd/*
srw-rw-rw- 1 root root 0 Aug 18 10:04 /var/run/pcscd/pcscd.comm
-rw-r--r-- 1 root root 6 Aug 18 10:04 /var/run/pcscd/pcscd.pid
doug@XUbuntu-20:/tmp$ ls -l /run/pcscd/*
srw-rw-rw- 1 root root 0 Aug 18 10:04 /run/pcscd/pcscd.comm
-rw-r--r-- 1 root root 6 Aug 18 10:04 /run/pcscd/pcscd.pid
```
and `cat /run/pcscd/pcscd.pid` does indeed have 6646

I do not have a good solution, but have been playing around with a test script that might be helpful 
[test-sudo.sh.txt](https://github.com/OpenSC/OpenSC/files/7008352/test-sudo.sh.txt)

Simplest solution for now would not try and restart and kill pcscd. 

Next is to use a sleep; and wait for /var/run/pcscd/pcscd.pid to be created. The script has a sleep 10.

**Comment by Jakuje** [comment]
Thank you for having a look into that.

I missed the sudo issue, but each of the tasks should run the start and stop only once in the particular task so we shold not get more pcscd's in one task.

I think in all the tests, we are introducing some module into the pcscd, which requires us to restart the pcscd before we can use the virtual card.

I gave it a try improving the setup, adding the cleanup to `trap` and waiting for the pid file before continuing, but nothing helped so far. I will have to play with it a bit more inside local container.

**Comment by dengert** [comment]
I too have been looking at it all morning with a test script. 

If multiple scripts are restarting pcscd should we do it only once before ? Does git action try and run the scripts in parallel?

Could we use the running pcscd, But not sure where it is writing errors. if its syslog, could be grep for that later?

`sudo /usr/sbin/pcscd -f | sed -e 's/^/pcscd: /;' &`
the sed will not run until /usr/sbin/pcscd  exits. The -f says it writes to stderr, you need to add -e  or -d at least to test?

The message" 
`00000000 [140426313230272] utils.c:81:GetDaemonPid() Can't open /run/pcscd/pcscd.pid: No such file or directory` 
is coming from the pcscd we are starting. if I add -e and don't use systemctl to stop the running pcscd I get
```
00000000 [139664321689536] pcscdaemon.c:482:main() file /run/pcscd/pcscd.comm already exists.
00000289 [139664321689536] pcscdaemon.c:484:main() Another pcscd (pid: 10069) seems to be running.
```
and it exits.

**Comment by Jakuje** [comment]
The most curious problem is that it used to work 8 days ago just fine:

https://github.com/OpenSC/OpenSC/actions/runs/1119707449

I tried to compare the raw logs:

**Comment by dengert** [comment]
I notice that the scripts do "git clone" but no checkouts. So  in effect we are testing test the "master" branches of many different repros with the OpenSC PRs. 
 
Should we be checking out some stable versions based on tags of some of these repros?

How can I test a from my repro without having to summit a PR?

**Comment by Jakuje** [comment]
Yeah. But there are not that much of projects we build and they are not changing that frequently.

You can run the github actions in your fork, if you enable github actions there, I think. They are made to be executed on PRs (if some c code is chaning) or on any push to any branch.

I just play with it now locally in ubuntu container. It is easy to clone there the repo and execute separate scripts. But they do not have systemd so it requires some more tweaks.

**Comment by Jakuje** [comment]
When I tried to run the threads tests with softhsm [1], the first test still crashes and the second one hangs, both inside of softhsm.

https://github.com/OpenSC/OpenSC/pull/2381/commits/28d5df07cda7f42893658194d634b4be23c2c007

The ubuntu has ancient softhsm from 2018, so I will probably drop that commit so far. Instead, I would use our opensc pkcs11 library, but that will not have any slots in CI so it is going to be XFAIL until we figure out something better.


For the CI, I just ran a couple of commands to make the same scripts working in local container and all worked just fine so I assume there is something broken in the virtual environment of Github Actions, which is breaking the pcscd/vsmartcard interaction. Both the CAC and javacard tests fail, which go through this way, but the OSEID works as it has different plugin to pcscd.

At least with the changes in #2381, the CAC test no longer hangs for 6 hours.
